src/storage.rs

```diff
@@ -126,136 +172,164 @@ impl MemStorageCore {
         cs: ConfState,
         pending_membership_change: Option<(ConfState, u64)>,
     ) {
-        self.snapshot.mut_metadata().set_conf_state(cs);
+        self.raft_state.conf_state = cs;
         if let Some((cs, idx)) = pending_membership_change {
-            self.snapshot
-                .mut_metadata()
-                .set_pending_membership_change(cs);
-            self.snapshot
-                .mut_metadata()
-                .set_pending_membership_change_index(idx);
+            self.raft_state.pending_conf_state = Some(cs);
+            self.raft_state.pending_conf_state_start_index = Some(idx);
+        }
+    }
+
+    fn inner_first_index(&self) -> u64 {
```

Comment:
For these `inner_` methods, what is 'inner' about them? If they are just meant to be for internal use, then I think you can drop the `inner_` prefix and rely on the fact that they are not `pub`.